### PR TITLE
Stop passing around IServiceProvider in RazorPreCompiler

### LINKED ISSUE #2204: Stop passing around IServiceProvider in RazorPreCompiler
https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNet.Mvc.Razor/Razor/PreCompileViews/RazorPreCompiler.cs#L30

We should always pass explicit dependencies and not assume a populated ISP. It makes it hard to understand what dependencies are actually required.

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #2204

### CODE REVIEW

**Comment by davidfowl** [comment]
:shipit: !!! Do this with more things in general.

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilationOptionsProviderExtension.cs [hunk 1]
```diff
@@ -32,5 +32,24 @@ public static class CompilationOptionsProviderExtension
                                                               applicationEnvironment.Configuration)
                                           .ToCompilationSettings(applicationEnvironment.RuntimeFramework);
         }
+
+        /// <summary>
+        /// Parses the <see cref="ICompilerOptions"/> for a project specified by <see cref="IProjectContext"/>
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilationOptionsProviderExtension.cs]: `<see cref="IProjectContext"/>` -> `<paramref name="projectContext"/>`

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilationOptionsProviderExtension.cs [hunk 2]
```diff
@@ -32,5 +32,24 @@ public static class CompilationOptionsProviderExtension
...
+        /// and returns a <see cref="CompilationSettings"/> used for Roslyn compilation.
+        /// </summary>
+        /// <param name="compilerOptionsProvider">
+        /// A <see cref="ICompilerOptionsProvider"/> that reads compiler options.
+        /// </param>
+        /// <param name="projectContext">The <see cref="IProjectContext"/> for the application being compiled.</param>
+        /// <returns>The <see cref="CompilationSettings"/> for the application being compiled.</returns>
+        public static CompilationSettings GetCompilationSettings(
+            [NotNull] this ICompilerOptionsProvider compilerOptionsProvider,
+            [NotNull] IProjectContext projectContext)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilationOptionsProviderExtension.cs]: why is this a `public` extension method and not a `private static` method in `RazorPreCompileModule`?

File: src/Microsoft.AspNet.Mvc/RazorPreCompileModule.cs [hunk 3]
```diff
@@ -76,76 +64,5 @@ public virtual void BeforeCompile(IBeforeCompileContext context)
         public void AfterCompile(IAfterCompileContext context)
         {
         }
-
-        // REVIEW: Welcome back to life delegating SP!!!
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc/RazorPreCompileModule.cs]: :dancers:

File: src/Microsoft.AspNet.Mvc.Razor/Razor/PreCompileViews/RazorPreCompiler.cs [hunk 4]
```diff
@@ -60,6 +42,8 @@ public class RazorPreCompiler
         /// </summary>
         public bool GenerateSymbols { get; set; }
 
+        protected IFileProvider FileProvider { get; }
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/Razor/PreCompileViews/RazorPreCompiler.cs]: how does making this a property relate to the bug you're fixing?

**Comment by dougbu** [comment]
:watch: mostly for a couple of answers

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilationOptionsProviderExtension.cs (hunk 2 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilationOptionsProviderExtension.cs]: The use case was for users to be use these extensions in their custom precompilation modules - it was primarily the reason we have the other extension method.

File: src/Microsoft.AspNet.Mvc.Razor/Razor/PreCompileViews/RazorPreCompiler.cs (hunk 4 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Razor/PreCompileViews/RazorPreCompiler.cs]: Minor cleanup.

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilationOptionsProviderExtension.cs (hunk 2 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilationOptionsProviderExtension.cs]: if that's the case, better / more discoverable as `protected` methods in our base precompilation module

File: src/Microsoft.AspNet.Mvc.Razor/Razor/PreCompileViews/RazorPreCompiler.cs (hunk 4 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/Razor/PreCompileViews/RazorPreCompiler.cs]: :ok:

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilationOptionsProviderExtension.cs (hunk 2 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilationOptionsProviderExtension.cs]: Perhaps. When the code was written, the refactoring prevented doing this - I think it's in better shape now to allow this

**Comment by pranavkm** [comment]
Updated.

File: src/Microsoft.AspNet.Mvc/RazorPreCompileModule.cs [hunk 5]
```diff
@@ -77,75 +65,14 @@ public void AfterCompile(IAfterCompileContext context)
...
-            public FrameworkName RuntimeFramework
-            {
-                get
-                {
-                    return _originalApplicationEnvironment.RuntimeFramework;
-                }
-            }
+            return compilerOptionsProvider.GetCompilerOptions(projectContext.Name,
+                                                              projectContext.TargetFramework,
+                                                              projectContext.Configuration)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc/RazorPreCompileModule.cs]: nit: will be easier to maintain if you just hit <Enter> after the open paren

**Comment by dougbu** [comment]
:shipit:

**Comment by HaoK** [comment]
Yay, now I don't have to keep fixing this code whenever hosting changes :)

**Comment by davidfowl** [comment]
@pranavkm Can you get this in?

**Comment by pranavkm** [comment]
Sure. The bug was marked for beta5 - I was under the impression I had to wait until we branched. I'll get this in.

**Pull request closed without merging**